File: Patrones/Creacionales/Singleton/Java/src/BaseDeDatos.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;

public class BaseDeDatos {
    // Variable estática que almacena la única instancia de la clase
    private static BaseDeDatos instancia;

    private String name;

    // Constructor privado para evitar que se creen instancias desde fuera de la clase
    private BaseDeDatos() {
        this.name = "MiBaseDeDatos " + LocalDateTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
        System.out.println("Conectando a la base de datos...");
    }

    // Método público estático para obtener la única instancia de la clase
    public static BaseDeDatos obtenerInstancia() {
        if (instancia == null) {
            instancia = new BaseDeDatos();
        }
        return instancia;
    }

    // Un método de ejemplo para mostrar que la instancia es única
    public void ejecutarConsulta(String consulta) {
        System.out.println("Ejecutando consulta: " + consulta + " con BD: " + this.name);
    }
}

File: Patrones/Creacionales/Singleton/Python/BaseDeDatos.py
from datetime import datetime

class BaseDeDatos:
    # Variable de clase para almacenar la única instancia
    __instancia = None

    # Constructor privado para evitar instanciación directa
    def __init__(self):
        if BaseDeDatos.__instancia is not None:
            raise Exception("Esta clase es un Singleton. Usa obtener_instancia()")
        else:
            self.__name = f"MiBaseDeDatos {datetime.now().strftime("%H:%M:%S")}"
            print("Conectando a la base de datos...")

    # Método estático para obtener la única instancia de la clase
    @staticmethod
    def obtener_instancia():
        if BaseDeDatos.__instancia is None:
            BaseDeDatos.__instancia = BaseDeDatos()
        return BaseDeDatos.__instancia

    # Un método de ejemplo para demostrar el uso de la instancia única
    def ejecutar_consulta(self, consulta):
        print(f"Ejecutando consulta: {consulta} con BD: {self.__name}")
